lib/Dialect/FIRRTL/Ops.cpp

```diff
@@ -74,6 +74,24 @@ static LogicalResult verifyCircuitOp(CircuitOp &circuit) {
     return failure();
   }
 
+  // Check that no extmodule defname conflicts with another module
+  // name in the circuit.
+  for (auto &op : *circuit.getBody()) {
+    if (auto extModule = dyn_cast<FExtModuleOp>(op)) {
+      if (auto defname = extModule.defnameAttr()) {
+        auto collidingModule = circuit.lookupSymbol(defname.getValue());
+        if (collidingModule && isa<FModuleOp>(collidingModule)) {
+          op.emitOpError(
+                "attribute 'defname' with value '" + defname.getValue() +
+                "' conflicts with the name of another module in the circuit")
+                  .attachNote(collidingModule->getLoc())
+              << "previous module declared here";
```

Comment:
It's late, so I forget the semantics here, but is this ok?

```
extmodule @foo defname = "xyz" ...
extmodule @bar  defname = "xyz" ...
```

or is that a conflict?  If that is a conflict, then I think you should keep the symbol names in a map that is populated during a scan.  Dump both the module names into the set as well as the defname names.

Reply:
That might be a collision, but just having the same `defname` isn't enough to know. It requires also looking at whether the ports match and whether or not the `ExtModule`s have parameters. This all stems from the fact that FIRRTL IR can't represent parameterized modules. So, you need one `ExtModule` for each different instantiation of a Verilog module with specific parameters. Two `ExtModule`s may then be rightly referring to the same Verilog module or it may be user error.

How about if I get all the `ExtModule` checking into this one PR instead of breaking it up? It's just some more compartmentalized lines in `verifyCircuitOp`.

Reply:
I'm happy for you to make the call here - it is ok to land this as an incremental step, or hold it to iterate on.  Whatever you prefer

Reply:
I updated this to include the full checks to get it to pass the same tests as the Scala FIRRTL Compiler for extmodules.

Reply:
nice!